File: apps/server/src/routes/spans.rs
use actix_web::{web, HttpResponse};

use crate::auth::ApiActor;
use crate::db::DbPool;
use crate::error::AppResult;
#[cfg(feature = "openapi")]
use crate::models::SpanResponse;
use crate::pagination::{ListSpansQuery, OffsetPaginatedResponse};
use crate::services::access::{self, Action};
use crate::services::span::{SpanFilters, SpanService};

#[cfg(feature = "openapi")]
use utoipa::OpenApi;

#[cfg_attr(feature = "openapi", utoipa::path(
    get,
    path = "/api/projects/{project_id}/spans",
    tag = "Spans",
    params(
        ("project_id" = i32, Path, description = "Project ID"),
        ListSpansQuery,
    ),
    responses(
        (status = 200, description = "Paginated span list", body = inline(crate::pagination::OffsetPaginatedResponse<SpanResponse>)),
        (status = 401, description = "Unauthorized", body = crate::error::ErrorResponse),
        (status = 404, description = "Not found", body = crate::error::ErrorResponse),
    ),
    security(("bearer_auth" = [])),
))]
/// GET /api/projects/{project_id}/spans
/// Lists spans for a project with offset-based pagination (newest by
/// start_timestamp first), optionally filtered by op/status/trace_id.
/// Returns spans regardless of origin — standalone or extracted from a
/// transaction — since both are stored in the same table.
pub async fn list_spans(
    pool: web::Data<DbPool>,
    path: web::Path<i32>,
    query: web::Query<ListSpansQuery>,
    actor: ApiActor,
) -> AppResult<HttpResponse> {
    let project_id = path.into_inner();

    access::require(
        pool.get_ref(),
        actor.is_admin(),
        actor.user_id(),
        project_id,
        Action::ViewProject,
    )
    .await?;

    let page = query.page.max(1);
    let per_page = query.per_page.clamp(1, 100);

    let filters = SpanFilters {
        op: query.op.clone(),
        status: query.status.clone(),
        trace_id: query.trace_id.clone(),
        operation_type: query.operation_type.clone(),
    };

    let (spans, total_count) =
        SpanService::list_offset(pool.get_ref(), project_id, page, per_page, &filters).await?;

    Ok(HttpResponse::Ok().json(OffsetPaginatedResponse::new(
        spans,
        total_count,
        page,
        per_page,
    )))
}

#[cfg(feature = "openapi")]
#[derive(OpenApi)]
#[openapi(paths(list_spans))]
pub struct SpansApi;

/// Configure span routes
pub fn configure(cfg: &mut web::ServiceConfig) {
    cfg.service(
        web::scope("/api/projects/{project_id}/spans").route("", web::get().to(list_spans)),
    );
}

File: packages/client/src/resources/spans.ts
import { offsetPaginatedResponseSchema, spanSchema } from '../schemas/index.js';
import type {
  ListSpansOptions,
  OffsetPaginatedResponse,
  Span,
} from '../types/index.js';
import { BaseResource } from './base.js';

/**
 * Spans API resource — the shared `spans` table (standalone "span" envelope
 * items AND transaction-embedded spans both live here and share this
 * response shape).
 */
export class SpansResource extends BaseResource {
  /**
   * List spans for a project with offset-based pagination (newest first),
   * optionally filtered by op/status/trace_id/operation_type.
   */
  async list(
    projectId: number,
    options?: ListSpansOptions,
  ): Promise<OffsetPaginatedResponse<Span>> {
    const searchParams: Record<string, string> = {};

    if (options?.page) {
      searchParams.page = String(options.page);
    }
    if (options?.per_page) {
      searchParams.per_page = String(options.per_page);
    }
    if (options?.op) {
      searchParams.op = options.op;
    }
    if (options?.status) {
      searchParams.status = options.status;
    }
    if (options?.trace_id) {
      searchParams.trace_id = options.trace_id;
    }
    if (options?.operation_type) {
      searchParams.operation_type = options.operation_type;
    }

    const data = await this.http
      .get(`api/projects/${projectId}/spans`, { searchParams })
      .json();

    return this.validate(data, offsetPaginatedResponseSchema(spanSchema));
  }
}
